design/build-clusters.md

```diff
@@ -0,0 +1,140 @@
+	1. Cluster name
+	2. Status - cluster status
+	3. Queue name
+	4. Authorization (initial phase will support JWT with key and sign options)
+	5. SCM Organization
+	6. SCM Context
+	7. Managed (Screwdriver / External) 
+
+Multiple build cluster onboarding process doc 
+	TBD
+
+## Design
+
+Redis queue in buildClusters
+
+![build-clusters-design.png](diagrams/build-clusters-design.png)
+
+
+### Scheduler service will be responsible 
+	1. identify build cluster and queue information for a build and queue jobs in respective queues.
+
+	2. authorize queue worker from build cluster and allow only authorized jobs from queues wrto build cluster.
+
+### Authorization
+Initial phase, we will go with JWT + private and public key authorization. Token expiry will be passed as part of signOption. Periodically cycle private+public key and signOption which has the expiry interval, and this will be a manual step which needs to be co-ordinated between Build cluster admin and Screwdriver team.
+
+### Yaml 
+
+```yml
+shared:
+    environment:
+    NODE_ENV: test
+    settings:
+        email:
+    addresses: [test@email.com, test2@email.com]
+    statuses: [SUCCESS, FAILURE]
+    annotations:
+        buildCluster: iOS
+jobs:
+    main:
+        requires: [~pr, ~commit]
+        sourcePaths: ["src/app/", "screwdriver.yaml"]
+        image: node:6
+        steps:
+    - init: npm install
+    - test: npm test
+    publish:
+    requires: main
+    image: node:6
+    steps:
+        - publish: npm publish
+    ...
+```   
+ 
+### New table for build cluster details
+Table: `buildClusters`
+
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganization` | text(200) | no | no | no | |
+| `scmContext` | text(200) | no | no | no | |
+| `isActive` | boolean | no | no | no | *0-false or 1-true* |
+| `authKeyFileName` | text(100) | no | no | yes | *encrypt and store the public key filename* |
+| `signOptionFileName` | text(100) | no | no | yes | *encrypt and store signOption filename { issuer: , subject: , audience: , expires: , algorithm: }* |
+| `queue` | text(100) | no | no | no | |
+| `managedBy` | text(50) | no | no | no | cluster managed by *screwdriver or external* |
+
+
+### Cache 
+
+	Use Redis HMSET to cache all active records from `buildClusters` table. Redis should be updated for any changes in `buildClusters` table.
+		HMSET <cluster_name> authKeyName "filename" signOption "{value}" queue "value" managedBy "screwdriver"
+		HGET <cluster_name> authKeyName
+		HGET <cluster_name> signOption
+
+### Queue
+
+Redis queue (Master) - It will hold queues for all build clusters. 
+
+Redis queue (Build Cluster) - It will hold queue wrto the build cluster.
+
+
+### Below listed apis need to be built to manage the cluster details
+
+| Method | url |
+| --- | --- |
+| `PUT` | ` /v4/buildClusters { } ` |
```

Comment:
can you change the url style to follow our hapi format? example: https://github.com/screwdriver-cd/screwdriver/blob/master/design/build-cache.md